File: code/python/day1.py
import sys
from pathlib import Path


def is_top_carry(top_carries, compare_carry):
    return len(top_carries) < 3 or any([ carry  < compare_carry for carry in top_carries ])

def add_or_update_carry(top_carries, compare_carry):
    top_carries.append(compare_carry)
    sorted(top_carries)
    return sorted(top_carries)[-3:]

def run():
    if len(sys.argv) < 2:
        filepath = Path("default.txt")
    else:
        filepath = Path(sys.argv[1])

    top_carries = []
    sum_carries = []
    with filepath.open() as f:
        for line in f.readlines():
            line = line.strip()
            if not len(line):
                total_carries = sum(sum_carries)
                if is_top_carry(top_carries, total_carries):
                    top_carries = add_or_update_carry(top_carries, total_carries)
                sum_carries.clear()
            else:
                if line.isdigit():
                    sum_carries.append(int(line))
    print(f"top = {sum(top_carries)}")
    print(f"set = {top_carries}")

if __name__ == '__main__':
    run()

File: code/rust/src/day1.rs
use std::env;
use std::fs;

fn is_top_carry(top_carries: &Vec<u32>, compared_carry: u32) -> bool {
    top_carries.len() < 3 || top_carries.iter().any(|&v| v < compared_carry)
}
fn add_or_update_top_carry(top_carries: &Vec<u32>, add_carry: u32) -> Vec<u32> {
    let mut new_top_carries = top_carries.clone();
    new_top_carries.push(add_carry);
    new_top_carries.sort();
    new_top_carries.iter().rev().take(3).map(|&v| v).collect()
}

fn main() -> Result<(), String> {
    let args: Vec<String> = env::args().collect();
    let file_path = args.get(1).ok_or("It is necessary an argument")?;
    let contents =
        fs::read_to_string(file_path).map_err(|_| "it was not possible open the file")?;

    let mut current_carriers: Vec<u32> = Vec::new();
    let mut top_carries: Vec<u32> = Vec::new();
    for line in contents.lines() {
        if line.is_empty() {
            let sum_carries = current_carriers.iter().sum();

            if is_top_carry(&top_carries, sum_carries) {
                top_carries = add_or_update_top_carry(&top_carries, sum_carries);
            }
            current_carriers.clear();
        }
        if let Ok(number) = line.parse::<u32>() {
            current_carriers.push(number);
        }
    }
    println!("{:}", top_carries.iter().sum::<u32>());

    Ok(())
}
